nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetTCP.java

```diff
@@ -0,0 +1,668 @@
+    private ExecutorService executorService;
+    private transient ComponentLog log = getLogger();
+    private transient String originalServerAddressList;
+    private transient String backupServer;
+    private transient int batchSize;
+
+    /**
+     * Bounded queue of messages events from the socket.
+     */
+    protected final BlockingQueue<String> socketMessagesReceived = new ArrayBlockingQueue<>(256);
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new PropertyDescriptor.Builder()
+                .required(false)
+                .name(propertyDescriptorName)
+                .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+                .dynamic(true)
+                .expressionLanguageSupported(true)
+                .build();
+    }
+
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws ProcessException {
+
+
+        final int rcvBufferSize = context.getProperty(RECEIVE_BUFFER_SIZE).asInteger();
+        final boolean keepAlive = context.getProperty(KEEP_ALIVE).asBoolean();
+        final int connectionTimeout = context.getProperty(CONNECTION_TIMEOUT).asTimePeriod(TimeUnit.SECONDS).intValue();
+        final int connectionRetryCount = context.getProperty(CONNECTION_ATTEMPT_COUNT).asInteger();
+        originalServerAddressList = context.getProperty(ENDPOINT_LIST).getValue();
+        final String[] serverAddresses = originalServerAddressList.split(",");
+        backupServer = context.getProperty(FAILOVER_ENDPOINT).getValue();
+        executorService = Executors.newFixedThreadPool(serverAddresses.length);
+        batchSize = context.getProperty(BATCH_SIZE).asInteger();
+
+        for (final Map.Entry<PropertyDescriptor, String> entry : context.getProperties().entrySet()) {
+            final PropertyDescriptor descriptor = entry.getKey();
+            if (descriptor.isDynamic()) {
+                dynamicAttributes.put(descriptor.getName(), entry.getValue());
+
+            }
+        }
+
+        //Go through all of the servers that we need to connect to and establish a connection to each of them
+        //this will result in all of them generating data that goes into the flow files.
+        for (final String hostPortPair : serverAddresses) {
+            //split pair
+            final String[] hostAndPort = hostPortPair.split(":");
+
+
+            SocketRecveiverThread socketRecveiverThread = null;
+            try {
+                socketRecveiverThread = createSocketRecveiverThread(rcvBufferSize, keepAlive, connectionTimeout, connectionRetryCount, hostAndPort);
+            } catch (Exception ex) {
```

Comment:
This is pretty dangerous since any kind of blockage in _OnSchedule_ may result in TimeoutException after which _OnSchedule_ will be re-invoked essentially attempting to reconnect to things that may already being connected to and now failing with other types of exceptions or errors.
OnSchedule is generally used to set properties and perform other localized initialization routines. The logic of obtaining the connection should go into onTrigger() and in the case of multi-threaded invocation synchronized. You can look at AMQP as well as JMS processors fro an example.